File: dot_agents/security-hooks/prompt_submit.py
#!/usr/bin/env python3
"""
UserPromptSubmit hook: capture the working-tree baseline for L2 diff review.

Runs at the start of each user turn for both Claude Code and Codex CLI.
Saves a baseline SHA representing the working tree state at this moment so
that `stop_review.py` can compute "this turn's changes".

Strategy:
- `git stash create -u` produces a commit object capturing both tracked
  modifications AND untracked files, without touching the working tree.
- If the working tree is clean, fall back to HEAD.
- If not in a git repo, write nothing — stop_review.py falls back to HEAD.

Disable: SECURITY_GUIDANCE_DISABLE=1.

Output: nothing (silent capture). Always exits 0.
"""

from __future__ import annotations

import json
import shutil
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from lib.common import (  # noqa: E402
    MAX_FILE_BYTES,
    STATE_DIR,
    atomic_write_text,
    globally_disabled,
    log,
    read_payload,
    safe_session_key,
)


def baseline_path(session_id: str) -> Path:
    return STATE_DIR / "baseline" / f"{safe_session_key(session_id)}.json"


def untracked_snapshot_dir(session_id: str) -> Path:
    return STATE_DIR / "baseline" / f"{safe_session_key(session_id)}.untracked"


def snapshot_untracked_files(cwd: Path, session_id: str, untracked: list[str]) -> None:
    """
    Copy current contents of each untracked path to a per-session snapshot dir.
    stop_review.py uses this to detect modifications to files that were already
    untracked at baseline time (a case `git diff` cannot cover).

    Skips files larger than MAX_FILE_BYTES to bound disk usage.
    """
    snap_root = untracked_snapshot_dir(session_id)
    # Wipe any stale snapshot from a previous turn with the same session_id.
    if snap_root.exists():
        shutil.rmtree(snap_root, ignore_errors=True)
    for rel in untracked:
        src = cwd / rel
        try:
            if not src.is_file():
                continue
            if src.stat().st_size > MAX_FILE_BYTES:
                continue
        except OSError:
            continue
        dst = snap_root / rel
        try:
            dst.parent.mkdir(parents=True, exist_ok=True)
            dst.write_bytes(src.read_bytes())
        except Exception as e:
            log("prompt_submit", f"snapshot {rel} failed: {e}")


def capture(cwd: Path) -> dict | None:
    """Return baseline dict {sha, untracked} or None if not a git repo."""
    try:
        r = subprocess.run(
            ["git", "-C", str(cwd), "rev-parse", "--is-inside-work-tree"],
            capture_output=True, text=True, timeout=5,
        )
        if r.returncode != 0 or r.stdout.strip() != "true":
            return None
    except Exception as e:
        log("prompt_submit", f"git check failed: {e}")
        return None

    # Try stash create -u first. Returns empty stdout when tree is clean.
    sha = ""
    try:
        r = subprocess.run(
            ["git", "-C", str(cwd), "stash", "create", "-u"],
            capture_output=True, text=True, timeout=15,
        )
        if r.returncode == 0:
            sha = r.stdout.strip()
    except Exception as e:
        log("prompt_submit", f"stash create failed: {e}")

    if not sha:
        # Clean tree — use HEAD.
        try:
            r = subprocess.run(
                ["git", "-C", str(cwd), "rev-parse", "HEAD"],
                capture_output=True, text=True, timeout=5,
            )
            if r.returncode == 0:
                sha = r.stdout.strip()
        except Exception as e:
            log("prompt_submit", f"rev-parse HEAD failed: {e}")

    if not sha:
        return None

    # Record untracked file list at baseline time. New files created during
    # the turn won't appear in this list, which is how we identify them.
    untracked: list[str] = []
    try:
        r = subprocess.run(
            ["git", "-C", str(cwd), "ls-files", "--others", "--exclude-standard"],
            capture_output=True, text=True, timeout=10,
        )
        if r.returncode == 0:
            untracked = [ln for ln in r.stdout.splitlines() if ln]
    except Exception as e:
        log("prompt_submit", f"ls-files failed: {e}")

    return {"sha": sha, "untracked": untracked, "cwd": str(cwd)}


def main() -> int:
    if globally_disabled():
        return 0

    payload = read_payload()
    if not payload:
        return 0

    session_id = str(payload.get("session_id") or "default")
    cwd = Path(payload.get("cwd") or ".").resolve()

    baseline = capture(cwd)
    if baseline is None:
        return 0

    atomic_write_text(baseline_path(session_id), json.dumps(baseline))
    snapshot_untracked_files(cwd, session_id, baseline.get("untracked") or [])
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception as e:
        log("prompt_submit", f"unhandled: {e}")
        sys.exit(0)

File: dot_agents/security-hooks/prompt-submit.ts
#!/usr/bin/env bun
/**
 * UserPromptSubmit hook: capture the working-tree baseline for L2 diff review.
 *
 * Runs at the start of each user turn for both Claude Code and Codex CLI.
 * Saves a baseline SHA representing the working tree state at this moment so
 * that `stop_review.ts` can compute "this turn's changes".
 *
 * Strategy:
 * - `git stash create -u` produces a commit object capturing both tracked
 *   modifications AND untracked files, without touching the working tree.
 * - If the working tree is clean, fall back to HEAD.
 * - If not in a git repo, write nothing — stop_review.ts falls back to HEAD.
 *
 * Disable: SECURITY_GUIDANCE_DISABLE=1.
 *
 * Output: nothing (silent capture). Always exits 0.
 */

import { spawnSync } from 'node:child_process';
import {
  mkdirSync,
  readFileSync,
  rmSync,
  statSync,
  writeFileSync,
} from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import process from 'node:process';

import {
  MAX_FILE_BYTES,
  STATE_DIR,
  atomicWriteText,
  globallyDisabled,
  log,
  readPayload,
  safeSessionKey,
  stringifyError,
} from './lib/common.ts';

type Baseline = {
  sha: string;
  untracked: string[];
  cwd: string;
};

function baselinePath(sessionId: string): string {
  return join(STATE_DIR, 'baseline', `${safeSessionKey(sessionId)}.json`);
}

function untrackedSnapshotDir(sessionId: string): string {
  return join(STATE_DIR, 'baseline', `${safeSessionKey(sessionId)}.untracked`);
}

function runGit(
  cwd: string,
  args: string[],
  timeoutMs: number,
): {
  ok: boolean;
  stdout: string;
  stderr: string;
} {
  try {
    const r = spawnSync('git', ['-C', cwd, ...args], {
      encoding: 'utf8',
      timeout: timeoutMs,
    });
    return {
      ok: r.status === 0,
      stdout: r.stdout ?? '',
      stderr: r.stderr ?? '',
    };
  } catch (e) {
    return { ok: false, stdout: '', stderr: stringifyError(e) };
  }
}

function snapshotUntrackedFiles(
  cwd: string,
  sessionId: string,
  untracked: string[],
): void {
  const snapRoot = untrackedSnapshotDir(sessionId);
  try {
    rmSync(snapRoot, { recursive: true, force: true });
  } catch {
    /* ignore */
  }
  for (const rel of untracked) {
    const src = join(cwd, rel);
    try {
      const st = statSync(src);
      if (!st.isFile()) continue;
      if (st.size > MAX_FILE_BYTES) continue;
    } catch {
      continue;
    }
    const dst = join(snapRoot, rel);
    try {
      mkdirSync(dirname(dst), { recursive: true });
      writeFileSync(dst, readFileSync(src));
    } catch (e) {
      log('prompt-submit', `snapshot ${rel} failed: ${stringifyError(e)}`);
    }
  }
}

function capture(cwd: string): Baseline | null {
  const check = runGit(cwd, ['rev-parse', '--is-inside-work-tree'], 5000);
  if (!check.ok || check.stdout.trim() !== 'true') {
    if (!check.ok && check.stderr)
      log('prompt-submit', `git check failed: ${check.stderr.slice(0, 200)}`);
    return null;
  }

  let sha = '';
  const stash = runGit(cwd, ['stash', 'create', '-u'], 15_000);
  if (stash.ok) {
    sha = stash.stdout.trim();
  } else if (stash.stderr) {
    log('prompt-submit', `stash create failed: ${stash.stderr.slice(0, 200)}`);
  }

  if (!sha) {
    const head = runGit(cwd, ['rev-parse', 'HEAD'], 5000);
    if (head.ok) sha = head.stdout.trim();
    else if (head.stderr)
      log(
        'prompt-submit',
        `rev-parse HEAD failed: ${head.stderr.slice(0, 200)}`,
      );
  }

  if (!sha) return null;

  const ls = runGit(
    cwd,
    ['ls-files', '--others', '--exclude-standard'],
    10_000,
  );
  const untracked = ls.ok
    ? ls.stdout.split('\n').filter((ln) => ln.length > 0)
    : [];
  if (!ls.ok && ls.stderr)
    log('prompt-submit', `ls-files failed: ${ls.stderr.slice(0, 200)}`);

  return { sha, untracked, cwd };
}

async function main(): Promise<number> {
  if (globallyDisabled()) return 0;

  const payload = await readPayload();
  if (Object.keys(payload).length === 0) return 0;

  const sessionId = String(payload.session_id ?? 'default');
  const cwd = resolve(payload.cwd ?? '.');

  const baseline = capture(cwd);
  if (baseline === null) return 0;

  atomicWriteText(baselinePath(sessionId), JSON.stringify(baseline));
  snapshotUntrackedFiles(cwd, sessionId, baseline.untracked);
  return 0;
}

try {
  process.exit(await main());
} catch (e) {
  log('prompt-submit', `unhandled: ${stringifyError(e)}`);
  process.exit(0);
}
